### PR TITLE
Prevent Repository objects from being prematurely GCed

### LINKED ISSUE #1315: Find out why / when LibGit2Sharp is crashing
### This investigation is related to the following issue

GitHub extension crashes when viewing pull requests: #1306

- Crash appears in LibGit2Sharp version: 0.24, 0.25.0-preview-0033
- Crash not yet seen in LibGit2Sharp version: 0.23.0,  0.23.1

### How to reproduce

Compile and run the following code in `Release` configuration.

```
        static void Main(string[] args)
        {
            var path = @"C:\source\github.com\github\VisualStudio";
            for (int count = 0; count < 1000; count++)
            {
                var status = new Repository(path).RetrieveStatus();
                Console.WriteLine(count + ": " + status.IsDirty);
            }
        }
```

Eventually something like this will happen:

image

Here is another very common exception:
```
System.AccessViolationException was unhandled
  HResult=-2147467261
  Message=Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
  Source=LibGit2Sharp
  StackTrace:
       at LibGit2Sharp.Core.NativeMethods.git_status_list_new(git_status_list*& git_status_list, git_repository* repo, GitStatusOptions options)
       at LibGit2Sharp.Core.Proxy.git_status_list_new(RepositoryHandle repo, GitStatusOptions options) in C:\projects\libgit2sharp\LibGit2Sharp\Core\Proxy.cs:line 2931
       at LibGit2Sharp.RepositoryStatus..ctor(Repository repo, StatusOptions options) in C:\projects\libgit2sharp\LibGit2Sharp\RepositoryStatus.cs:line 60
       at LibGit2Sharp.RepositoryExtensions.RetrieveStatus(IRepository repository) in C:\projects\libgit2sharp\LibGit2Sharp\RepositoryExtensions.cs:line 696
       at Libgit2Repro.Program.Main(String[] args) in c:\users\passp\onedrive\documents\visual studio 2015\Projects\Libgit2Repro\Libgit2Repro\Program.cs:line 20
       at System.AppDomain._nExecuteAssembly(RuntimeAssembly assembly, String[] args)
       at System.AppDomain.ExecuteAssembly(String assemblyFile, Evidence assemblySecurity, String[] args)
       at Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssembly()
       at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
       at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
       at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
       at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
       at System.Threading.ThreadHelper.ThreadStart()
```

This only seems to happen when compiled in `Release` configuration.  When compiled in `Debug` configuration, it can run all 1000 iterations and never crash.

### PR DESCRIPTION
**Pull request by jcansdale**
## The Bug

The following code can fail on the `repo.RetrieveStatus()` line with an `AccessViolationException`:

```c#
var repo = new Repository(path);
var status = repo.RetrieveStatus();
```

This is because `RetrieveStatus` creates a native pointer to the `Repository` object and but doesn't hold on to a managed reference. This allows the `Repository` object to be GCed before `RetrieveStatus` has finished with the native pointer.

- When does an object become available for garbage collection?https://blogs.msdn.microsoft.com/oldnewthing/20100810-00/?p=13193/

To vaoid random crashes, we need to be careful to keep a reference to `Repository` when calling its methods. The above code can be fixed like this:

```c#
var repo = new Repository(path);
var status = repo.RetrieveStatus();
GC.KeepAlive(repo);
```

Alternatively since the `Repository` object implements `IDisposable`, the following will keep a reference and dispose of it  afterwards:

```c#
using (var repo = new Repository(path))
{
    var status = repo.RetrieveStatus();
}
```

This PR wraps all temporary references to a `Repository` object in a `using` statement.

## Possible Gotchas

- This bug only shows up when the extension is compiled in `Release` configuration (in `Debug` configuration, references are held on to for longer and this bug doesn't occur).
- We need to make sure we don't return objects that keep references to disposed `Repository` objects. For example here, we need to make sure that `CommitMessage` doesn't cache a reference to the disposed `Repository` object:
https://github.com/github/VisualStudio/blob/7fb09a3282369eeb944b3b3b5d1b1c45c5d26c3f/src/GitHub.App/Services/PullRequestService.cs#L101

Fixes #1315
See also: https://github.com/libgit2/libgit2sharp/issues/1513

### CODE REVIEW

**Comment by jcansdale** [comment]
I got the following when trying to create a new PR:

image

```
 	LibGit2Sharp.dll!LibGit2Sharp.Core.Proxy.git_revparse_ext(LibGit2Sharp.Core.Handles.RepositoryHandle repo, string objectish)	Unknown
 	LibGit2Sharp.dll!LibGit2Sharp.Core.Proxy.git_revparse_single(LibGit2Sharp.Core.Handles.RepositoryHandle repo, string objectish)	Unknown
 	LibGit2Sharp.dll!LibGit2Sharp.Repository.Lookup(string objectish, LibGit2Sharp.Core.GitObjectType type, LibGit2Sharp.Core.LookUpOptions lookUpOptions)	Unknown
 	LibGit2Sharp.dll!LibGit2Sharp.Repository.Lookup(string objectish, LibGit2Sharp.ObjectType type)	Unknown
 	LibGit2Sharp.dll!LibGit2Sharp.RepositoryExtensions.Lookup<LibGit2Sharp.Commit>(LibGit2Sharp.IRepository repository, string objectish)	Unknown
>	GitHub.App.dll!GitHub.Services.GitClient.GetMessagesForUniqueCommits.AnonymousMethod__0() Line 486	C#
```

File: src/GitHub.App/Services/PullRequestService.cs [hunk 1]
```diff
@@ -95,8 +95,11 @@ public IObservable<string> GetPullRequestTemplate(ILocalRepositoryModel reposito
         {
             return Observable.Defer(() =>
             {
-                var repo = gitService.GetRepository(repository.LocalPath);
-                return gitClient.GetMessagesForUniqueCommits(repo, baseBranch, compareBranch, maxCommits).ToObservable();
+                // CommitMessage doesn't keep a reference to Repository
+                using (var repo = gitService.GetRepository(repository.LocalPath))
+                {
+                    return gitClient.GetMessagesForUniqueCommits(repo, baseBranch, compareBranch, maxCommits).ToObservable();
```
Comment (jcansdale) [inline src/GitHub.App/Services/PullRequestService.cs]: ToObservable / threading issue.

File: src/GitHub.App/Services/PullRequestService.cs [hunk 2]
```diff
@@ -203,58 +217,69 @@ public IObservable<BranchTrackingDetails> CalculateHistoryDivergence(ILocalRepos
...
             return Observable.Defer(() =>
             {
-                var repo = gitService.GetRepository(repository.LocalPath);
-                var result = GetLocalBranchesInternal(repository, repo, pullRequest).Select(x => new BranchModel(x, repository));
-                return result.ToObservable();
+                // BranchModel doesn't keep a reference to repo
+                using (var repo = gitService.GetRepository(repository.LocalPath))
+                {
+                    var result = GetLocalBranchesInternal(repository, repo, pullRequest).Select(x => new BranchModel(x, repository));
+                    return result.ToObservable();
```
Comment (jcansdale) [inline src/GitHub.App/Services/PullRequestService.cs]: @grokys I have a bad feeling about this one. What do you think?

Comment (jcansdale) [reply src/GitHub.App/Services/PullRequestService.cs]: Yup, as expected!
image

Comment (grokys) [reply src/GitHub.App/Services/PullRequestService.cs]: Hmm yeah, should either call `ToList()` on result here.

For reference, the other option is to  destroy the `repo` with a [`Using`](http://www.introtorx.com/content/v1.0.10621.0/11_AdvancedErrorHandling.html#Using) operator, I don't think that'd be necessary here however.

**Comment by jcansdale** [comment]
Hmmm.

image

**Comment by jcansdale** [comment]
I'm seeing an issue where after clicking the checkout link, the checkout appears to succeed but the UI doesn't update. Any idea what might be causing that?

**Comment by grokys** [comment]
> I'm seeing an issue where after clicking the checkout link, the checkout appears to succeed but the UI doesn't update. Any idea what might be causing that?

Not seen that before. Does it happen on master?

**Comment by jcansdale** [comment]
@grokys I thought this was likely an issue introduced with this PR.  Haven't seen it on master before.

File: test/GitHub.InlineReviews.UnitTests/TestDoubles/FakeDiffService.cs [hunk 3]
```diff
@@ -54,7 +54,9 @@ public string AddFile(string path, string contents, string commitAlias)
         public void Dispose()
         {
             var path = repository.Info.WorkingDirectory;
-            repository.Dispose();
+
+            // NOTE: IDiffService doesn't own the Repository object
+            //repository.Dispose();
```
Comment (grokys) [inline test/GitHub.InlineReviews.UnitTests/TestDoubles/FakeDiffService.cs]: I'd rather remove this commented-out line.

**Pull request merged**